codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -726,31 +727,85 @@ private void writeHttpBindingMember(
+                                writer.write("if encoder.HasQuery(qkey) { continue }");
+                                writeQueryBinding(context, queryMapKeyMemberShape, queryMapValueTargetShape,
+                                        "qvalue", location, "qkey", "encoder", true);
+                            });
+                            break;
+
                         default:
                             throw new CodegenException("unexpected http binding found");
                     }
                 });
     }
 
+    /**
+     * Writes query bindings, as per the target shape. This method is shared
+     * between members modeled with Location.Query and Location.QueryParams.
+     * Precedence across Location.Query and Location.QueryParams is handled
+     * outside the scope of this function.
+     *
+     * @param context       is the generation context
+     * @param memberShape   is the member shape for which query is serialized
+     * @param targetShape   is the target shape of the query member.
+     *                      This can either be string, or a list/set of string.
+     * @param operand       is the member value accessor .
+     * @param location      is the location of the member - can be Location.Query
+     *                      or Location.QueryParams.
+     * @param locationName  is the key for which query is encoded.
+     * @param dest          is the query encoder destination.
+     * @param isQueryParams boolean representing if Location used for query binding is
+     *                      QUERY_PARAMS.
+     */
+    private void writeQueryBinding(
+            GenerationContext context,
+            MemberShape memberShape,
+            Shape targetShape,
+            String operand,
+            HttpBinding.Location location,
+            String locationName,
+            String dest,
+            boolean isQueryParams
+    ) {
+        GoWriter writer = context.getWriter();
+
+        if (targetShape instanceof CollectionShape) {
+            MemberShape collectionMember = CodegenUtils.expectCollectionShape(targetShape)
+                    .getMember();
+            writer.openBlock("for i := range $L {", "}", operand, () -> {
+                GoValueAccessUtils.writeIfZeroValue(context.getModel(), writer, collectionMember,
+                        operand + "[i]", () -> {
+                            writer.write("continue");
+                        });
+                writeHttpBindingSetter(context, writer, collectionMember, location, operand + "[i]",
+                        (w, s) -> {
+                            if (isQueryParams) {
+                                w.writeInline("$L.AddQuery($L).$L", dest, locationName, s);
+                            } else {
+                                w.writeInline("$L.AddQuery($S).$L", dest, locationName, s);
+                            }
+                        });
```

Comment:
You could simplify this a bit by creating the format string at the top of the if scope, like so:

```java
String addQuery = String.format("$L.AddQuery(%s).$L", isQueryParams ? "$L" : "$S");
```

Then you can just inline this:

```java
(w, s) -> w.writeInline(addQuery, dest, locationName, s));
```

The same sort of thing can be done for the non-collection binding block.